### PR TITLE
Reproduce Android

### PR DESCRIPTION
**Pull request by tanin47**
In order to support Android, here are the changes:
- Add a new reproducer for Android.
- The testcase supports the `files` field which specify the location and the content of necessary files on Android.
- Filtering logs from Android is copied from ClusterFuzz.

Symbolization hasn't been implemented yet. This works with: 5971973158207488

### CODE REVIEW

**Comment by inferno-chromium** [comment]
Oliver, can you take a first pass at this review.

**Comment by tanin47** [comment]
Gentle ping

File: tool/clusterfuzz/android.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+
+def adb(arg, **kwargs):
+  """Run adb with arg."""
+  return common.execute('adb', arg, cwd='.', **kwargs)
+
+
+def adb_shell(arg, **kwargs):
+  """Run adb shell with arg."""
+  return adb('shell "%s"' % arg, **kwargs)
```
Comment (oliverchang) [inline tool/clusterfuzz/android.py]: we should probably be more careful with escaping characters here since the input here is a shell command?

File: tool/clusterfuzz/android.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+
+def write_content(path, content):
+  """Write content to path on Android."""
+  with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+    print tmp_file
```
Comment (oliverchang) [inline tool/clusterfuzz/android.py]: Are these prints intended to be left here?

File: tool/clusterfuzz/reproducers.py [hunk 3]
```diff
@@ -546,3 +566,67 @@ def reproduce_crash(self):
...
+    """Reproduce crash on Android."""
+    android.reset(self.get_package_name())
+    android.ensure_active()
+    android.clear_log()
+
+    ret_value, _ = android.adb_shell(
+        'am start -a android.intent.action.MAIN '
+        "-n %s/com.google.android.apps.chrome.Main '%s'" %
+        (self.get_package_name(), self.get_testcase_path()))
+    time.sleep(30)
```
Comment (oliverchang) [inline tool/clusterfuzz/reproducers.py]: maybe put this in a constant somewhere?

File: tool/clusterfuzz/android.py (hunk 2 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Oops. thanks for pointing it out.

File: tool/clusterfuzz/reproducers.py (hunk 3 above)
Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: Done

File: tool/clusterfuzz/android.py (hunk 1 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  logger.info(
+      common.colorize(
+          '\nWriting %s:\n%s\n', common.BASH_GREEN_MARKER),
+      path, content)
+
+  adb('push %s %s' % (tmp_file.name, path))
+  common.delete_if_exists(tmp_file.name)
+
+
+def file_exist(path):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: This is just wrong, please see adb.fileExists

Some adb commands can emit text like Warning:, so we need to check return code.

File: tool/clusterfuzz/android.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+logger = logging.getLogger('clusterfuzz')
+
+
+def adb(arg, **kwargs):
+  """Run adb with arg."""
+  return common.execute('adb', arg, cwd='.', **kwargs)
+
+
+def adb_shell(arg, **kwargs):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Please use code from runADBShellCommand

  def _escape_specials(command):
    return command.replace('\\', '\\\\').replace('"', '\\"')

  if isinstance(cmd, list):
    cmd = ' '.join([str(i) for i in cmd])

  if cmd[0] not in ['"', "'"]:
    cmd = '"{}"'.format(_escape_specials(cmd))

File: tool/clusterfuzz/android.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+def file_exist(path):
+  """Check if the path exists on Android."""
+  _, output = adb_shell('ls %s 2> /dev/null' % path, exit_on_error=False)
+  return bool(output.strip())
+
+
+def ensure_asan(source_dir_path, device_id):
+  """Ensures ASan is installed on Android."""
+  if file_exist('/system/lib/libclang_rt.asan-arm-android.so'):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: This file is different for 32-bit and 64-bit asan. please don't assume it, also that file can be updated from time to time. so, running it everytime is fast, it will skip if the checksums of changes are different, so run it everytime.

File: tool/clusterfuzz/android.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    logger.info(
+        'The device has ASAN. Skip installing ASAN.')
+    return
+
+  logger.info(
+      'Installing ASAN into the Android device. The device will be restarted.')
+
+  common.check_confirm(
+      'Are you sure you want to install ASAN on the Android device %s?' %
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: You can say like this testcase needs to install ASAN on your device.

File: tool/clusterfuzz/android.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+      device_id)
+
+  common.execute(
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Are you running tools/clang/update.py at any step. this will make sure clang is updated locally.

File: tool/clusterfuzz/android.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+  """Convert Android's crash stack line into sanitizer_format."""
+  m_crash_state = re.match(
+      r'\s*#([0-9]+)\s+pc\s+([xX0-9a-fA-F]+)\s+(.+)', line)
+  if not m_crash_state:
+    return None
+
+  frame_no = int(m_crash_state.group(1))
+  frame_address = m_crash_state.group(2)
+  frame_binary = m_crash_state.group(3).strip()
+  # Ignore invalid frames, helps to prevent errors
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: new line before this line.

File: tool/clusterfuzz/android.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  # while symbolizing.
+  if '<unknown>' in frame_binary:
+    return
+  # Normalize frame address.
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: new line before this line.

File: tool/clusterfuzz/android.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    m_line = re.match('[^D]/([^:]+)[:] (.*)', line)
+    if not m_line:
+      continue
+
+    header = m_line.group(1)
+    content = m_line.group(2)
+
+    process_id, process_name = get_process_id_and_name(header)
+    if process_id != last_process_id:
+      filtered_output += '--------- %s (%d):\n' % (process_name, process_id)
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Isn't this to be done after process android and chrome stack lines. ? check original code in clusterfuzz.

File: tool/clusterfuzz/android.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+  adb_shell(
+      'input keyevent KEYCODE_POWER', print_command=False, print_output=False)
+
+  # Unlock.
+  adb_shell(
+      'input keyevent KEYCODE_MENU', print_command=False, print_output=False)
+  time.sleep(1)
+
+
+def set_content(table, key, value):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: s/set_content/set_content_setting as a less confusing name.

File: tool/clusterfuzz/android.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+      redirect_stderr_to_stdout=True,
+      stdout_transformer=output_transformer.Identity())
+  return output
+
+
+def kill(package_name):
+  """Kills the process with the package name."""
+  adb_shell('am force-stop %s' % package_name, exit_on_error=False)
+
+  # Click OK to close 'Unfortunately, Chrome has stopped'.
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Add a FIXME, this is very hacky.

File: tool/clusterfuzz/android.py (hunk 8 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Yes, the base class of binary providers run `tools/clang/update.py` for every build.

File: tool/clusterfuzz/android.py (hunk 9 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done

File: tool/clusterfuzz/android.py (hunk 10 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done

File: error/error/error.py [hunk 14]
```diff
@@ -294,3 +294,40 @@ def __init__(self, count, crash_signatures):
...
+
+  def __init__(self):
+    super(WaitForAndroidAfterInstallingAsanError, self).__init__(
+        message=self.MESSAGE, exit_code=self.EXIT_CODE)
+
+
+class NoAndroidDeviceIdError(ExpectedException):
+  """An exception is raised after installing ASAN on Android"""
+
+  MESSAGE = 'Please set the target Android device ID as the env {env_name}.'
```
Comment (tanin47) [inline error/error/error.py]: The message is only for this type of error. Every error has its own `MESSAGE`.

File: tool/clusterfuzz/android.py (hunk 11 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: In ClusterFuzz, it's processed *after*. But its result is added *before*. Other than that, their processing are independent (e.g. one uses header, the other uses body).

File: tool/clusterfuzz/android.py (hunk 12 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py (hunk 13 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done

File: tool/clusterfuzz/android.py (hunk 6 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py (hunk 7 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py (hunk 4 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Anecdotally, `test -f` doesn't seem to work well. For example:

```
$ adb shell "ls /system/lib/libvixl.so"
/system/lib/libvixl.so
$ adb shell "test -f /system/lib/libvixl.so; echo $?"
0
$ adb shell "test -f /system/lib/libvixlssss.so; echo $?"
0
```

If there's a warning, how about we check if the `path` is in `output` (after splitting by lines)?

File: tool/clusterfuzz/android.py (hunk 5 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+  with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+    tmp_file.write(content)
+
+  logger.info(
+      common.colorize(
+          '\nWriting %s:\n%s\n', common.BASH_GREEN_MARKER),
+      path, content)
+
+  adb('push %s %s' % (tmp_file.name, path))
+  common.delete_if_exists(tmp_file.name)
```
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done. I don't think we need remount here because we only push to /sdcard folder.

**Comment by tanin47** [comment]
Thank you for the comments. It's ready for another pass.

File: error/error/error.py (hunk 14 above)
Comment (inferno-chromium) [inline error/error/error.py]: Sorry i meant for {env_name}, that shouldn't change.

File: tool/clusterfuzz/android.py (hunk 4 above)
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Lets discuss this tmrw, i definitely remember it worked before.

File: error/error/error.py [hunk 16]
```diff
@@ -294,3 +294,40 @@ def __init__(self, count, crash_signatures):
...
+    super(GdbNotSupportedOnAndroidError, self).__init__(
+        message=self.MESSAGE, exit_code=self.EXIT_CODE)
+
+
+class WaitForAndroidAfterInstallingAsanError(ExpectedException):
+  """An exception is raised after installing ASAN on Android"""
+
+  MESSAGE = (
+      'Please wait until the Android device is rebooted and run the tool '
+      "again. If there's permission error, try `adb remount` and run the "
```
Comment (inferno-chromium) [inline error/error/error.py]: s/try `adb remount`/make sure that `adb remount` succeeds

File: tool/clusterfuzz/android.py [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+
+SCREEN_LOCK_SEARCH_STRING = 'mShowingLockscreen=true'
+ASAN_DEVICE_SETUP_SUCCESS_SEARCH_STRING = (
+    'Please wait until the device restarts')
+
+logger = logging.getLogger('clusterfuzz')
+
+
+def adb(command, **kwargs):
+  """Run adb with arg."""
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: fix description for both function to have command.

File: tool/clusterfuzz/android.py [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+      common.colorize(
+          '\nWriting %s:\n%s\n', common.BASH_GREEN_MARKER),
+      path, content)
+
+  adb('push %s %s' % (tmp_file.name, path))
+  adb_shell('chmod 0644 %s' % path)
+  common.delete_if_exists(tmp_file.name)
+
+
+def file_exist(path):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: s/file_exist/file_exists ?

File: tool/clusterfuzz/android.py [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+  """Check if the path exists on Android."""
+  _, output = adb_shell('ls %s 2> /dev/null' % path, exit_on_error=False)
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Are we assuming that this is an absolute_path, if yes, lets rename vars.

File: tool/clusterfuzz/android.py [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+              'lib', 'clang', '*', 'lib', 'linux'),
+          device_id),
+      env={'ADB_PATH': common.check_binary('adb', cwd='.')},
+      redirect_stderr_to_stdout=True,
+      cwd='.')
+
+  # tool/clusterfuzz/resources/asan_device_setup.sh prints the below string
+  # on screen. We've decided to do this hack in order to avoid modifying
+  # asan_device_setup.sh which is copied from Chromium's src.
+  if ASAN_DEVICE_SETUP_SUCCESS_SEARCH_STRING in output:
```
Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: This is very very hacky, doesn't script return exit code. there are 2 cases, one where it changes stuff on device and exits with success, 2nd case where it does not need to modify anything and returns with success. otherwise it fails.

File: tool/clusterfuzz/reproducers.py [hunk 21]
```diff
@@ -546,3 +566,67 @@ def reproduce_crash(self):
...
+
+    if not device_id:
+      raise error.NoAndroidDeviceIdError(ANDROID_SERIAL_ENV)
+
+    return device_id
+
+  @common.memoize
+  def get_package_name(self):
+    """Get package name."""
+    return 'org.chromium.chrome'
```
Comment (inferno-chromium) [inline tool/clusterfuzz/reproducers.py]: This seems wrong, we use com.google.android.apps.chrome in job type.

File: tool/clusterfuzz/reproducers.py [hunk 22]
```diff
@@ -546,3 +566,67 @@ def reproduce_crash(self):
...
+    """Get the android device."""
+    device_id = os.environ.get(ANDROID_SERIAL_ENV)
+
+    if not device_id:
+      raise error.NoAndroidDeviceIdError(ANDROID_SERIAL_ENV)
+
+    return device_id
+
+  @common.memoize
+  def get_package_name(self):
```
Comment (inferno-chromium) [inline tool/clusterfuzz/reproducers.py]: I think like get_package_name,  line 614-615, 626-627, etc are chrome specific stuff which can go in AndroidChromeReproducer inherited class. Or just change this class name.

File: tool/clusterfuzz/reproducers.py [hunk 23]
```diff
@@ -546,3 +566,67 @@ def reproduce_crash(self):
...
+
+  def reproduce_crash(self):
+    """Reproduce crash on Android."""
+    android.reset(self.get_package_name())
+    android.ensure_active()
+    android.clear_log()
+
+    ret_value, _ = android.adb_shell(
+        'am start -a android.intent.action.MAIN '
+        "-n %s/com.google.android.apps.chrome.Main '%s'" %
```
Comment (inferno-chromium) [inline tool/clusterfuzz/reproducers.py]: This wont work for webview shell job type, why do we need to hardcode this, why don't we get this from job environment

File: tool/clusterfuzz/reproducers.py [hunk 24]
```diff
@@ -546,3 +566,67 @@ def reproduce_crash(self):
...
+    android.ensure_asan(
+        source_dir_path=self.binary_provider.get_source_dir_path(),
+        device_id=device_id)
+
+    super(AndroidReproducer, self).pre_build_steps()
+
+    for path, content in self.testcase.files.iteritems():
+      android.write_content(path, content)
+    android.write_content(
+        '/data/local/chrome-command-line', 'chrome %s' % self.args)
```
Comment (inferno-chromium) [inline tool/clusterfuzz/reproducers.py]: Good to get this from job environment COMMAND_LINE_PATH ?

File: error/error/error.py (hunk 16 above)
Comment (tanin47) [inline error/error/error.py]: Done.

File: tool/clusterfuzz/android.py (hunk 17 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Done.

File: tool/clusterfuzz/android.py (hunk 18 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: Ok, funnily enough, since we remove the asan check (e.g. checking if `/system/lib/...asan...` exist, we don't actually use this method anymore..

File: tool/clusterfuzz/android.py (hunk 19 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: The method is removed... we don't use it anymore after removing the asan check.

File: tool/clusterfuzz/android.py (hunk 20 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: The script succeeding (exiting with 0) means it either installs ASAN successfully or skips the installation.

Sadly, the script doesn't exit with different codes (https://github.com/google/clusterfuzz-tools/pull/428/files#diff-d224b682b1ec5006b23abcefaf30b29cR459). Here's the relevant code:

```
+    echo '>> Please wait until the device restarts'
 +else
 +    echo '>> Device is up to date'
 +fi
```

We need to distinguish between these 2 cases because we need stop if ASAN installation previously occurred.

Comment (inferno-chromium) [inline tool/clusterfuzz/android.py]: Isn't "Device is up to date" also a success.

File: tool/clusterfuzz/reproducers.py (hunk 24 above)
Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: I switched to reading from the stacktrace instead.

File: tool/clusterfuzz/reproducers.py (hunk 21 above)
Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: I think this is because we aren't building with Clank. We build from Chromium directly. Maybe that's a mistake. I'll change it to Clank.

Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: Actually, this involves changing binary_providers. Let's do that in a separate CL.

File: tool/clusterfuzz/reproducers.py (hunk 23 above)
Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: I'll fix this in a separate CL. We'll switch to Clank and get the command from the stacktrace.

Currently, the package name when building from Chrome is different. That's why this is hardcoded.

File: tool/clusterfuzz/android.py (hunk 20 above)
Comment (tanin47) [inline tool/clusterfuzz/android.py]: I've changed the name to `ASAN_BEING_INSTALLED_SEARCH_STRING`.

File: tool/clusterfuzz/reproducers.py (hunk 22 above)
Comment (tanin47) [inline tool/clusterfuzz/reproducers.py]: Done.

File: error/error/error.py (hunk 14 above)
Comment (tanin47) [inline error/error/error.py]: I see what you mean now. The main issue here is that I wouldn't like to hardcode the string `ANDROID_DEVICE_ID`. It's defined in `reproducers.ANDROID_DEVICE_ID_ENV`; it's also a bit better to let consumer of this class define the variable name.

**Comment by tanin47** [comment]
I've fixed and addressed all the comments. It's ready for another pass. Thank you.

File: error/error/error.py (hunk 14 above)
Comment (inferno-chromium) [inline error/error/error.py]: Ok.
